### DISCUSSION TITLE
How to get this working with VS Code

### DISCUSSION THREAD

**Discussion by JackUkleja** [microsoft/CsWin32#218] [General]
I've just added the package to a vanilla class library in latest VS Code. I'm a little rusty with this stuff, but I can't for the life of me figure out how to get the code generation working. I've followed the instructions with the `NativeMethods.txt` but nothing happens on build. The animated gif indicates it should generate some code.

**Comment by AArnott** [comment]
What do you mean by "nothing happens on build"? VS Code doesn't build code. The build happens with the .NET CLI, and that works:

![image](https://user-images.githubusercontent.com/3548/111705635-a71fea00-8806-11eb-8fc9-0a5806dcc704.png)

What I see failing isn't the build, but that Omnisharp, which provides the C# language service experience in VS Code, isn't running the code generation in memory, such that VS Code shows squiggles and "Problems" as if the code gen hadn't happened, although the build actually works.

I don't know what it would take for VS Code to take source generators into account. That is probably a good question for the Omnisharp folks. As I don't know anyone who works on that, I wonder if @jinujoseph has any ideas.

**Reply by jinujoseph** [reply]
@jasonmalinowski to take a look

**Reply by jasonmalinowski** [reply]
VS Code should be recognizing the generator here and running it, so something not working. Is there anything being logged to the Omnisharp output in the logs there?

**Reply by AArnott** [reply]
I don't see any errors in the omnisharp log.

<details>

```
Starting OmniSharp server at 3/18/2021, 5:09:56 PM
    Target: c:\Users\andarno\source\repos\CsWin32Sandbox\CsWin32Sandbox.sln

OmniSharp server started.
    Path: c:\Users\andarno\.vscode\extensions\ms-dotnettools.csharp-1.23.9\.omnisharp\1.37.6\OmniSharp.exe
    PID: 55928

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Windows 6.2.9200.0 (x64)
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 6 MSBuild instance(s)
            1: Visual Studio Community 2019 16.10.31104.28 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\VSonNet6\MSBuild\Current\Bin"
            2: Visual Studio Enterprise 2019 16.10.31115.18 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin"
            3: Visual Studio Enterprise 17 17.0.31112.230 - "C:\Program Files\Microsoft Visual Studio\17\Int.Preview\MSBuild\Current\Bin"
            4: Visual Studio Build Tools 2017 15.9.28307.1401 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin"
            5: Visual Studio Enterprise 2019 16.10.31115.58 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\main\MSBuild\Current\Bin"
            6: StandAlone 16.8.0 - "c:\Users\andarno\.vscode\extensions\ms-dotnettools.csharp-1.23.9\.omnisharp\1.37.6\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Enterprise 17 17.0.31112.230 - "C:\Program Files\Microsoft Visual Studio\17\Int.Preview\MSBuild\Current\Bin"
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.Completion.CompletionOptionsProvider, Order: 0
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.ImplementTypeWorkspaceOptionsProvider, Order: 110
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.BlockStructureWorkspaceOptionsProvider, Order: 140
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in 'c:\Users\andarno\source\repos\CsWin32Sandbox'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Could not find any Cake files
[info]: OmniSharp.MSBuild.ProjectSystem
        Detecting projects in 'c:\Users\andarno\source\repos\CsWin32Sandbox\CsWin32Sandbox.sln'.
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\andarno\source\repos\CsWin32Sandbox\CsWin32Sandbox\CsWin32Sandbox.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\andarno\source\repos\CsWin32Sandbox\ClassLibrary1\ClassLibrary1.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in 'c:\Users\andarno\source\repos\CsWin32Sandbox'.
[info]: OmniSharp.Script.ScriptProjectSystem
        Could not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location 'c:\Users\andarno\source\repos\CsWin32Sandbox' on host 8364.
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Users\andarno\source\repos\CsWin32Sandbox\CsWin32Sandbox\CsWin32Sandbox.csproj
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file 'c:\Users\andarno\source\repos\CsWin32Sandbox\CsWin32Sandbox\CsWin32Sandbox.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project 'c:\Users\andarno\source\repos\CsWin32Sandbox\CsWin32Sandbox\CsWin32Sandbox.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Users\andarno\source\repos\CsWin32Sandbox\ClassLibrary1\ClassLibrary1.csproj
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file 'c:\Users\andarno\source\repos\CsWin32Sandbox\ClassLibrary1\ClassLibrary1.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project 'c:\Users\andarno\source\repos\CsWin32Sandbox\ClassLibrary1\ClassLibrary1.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: CsWin32Sandbox
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: ClassLibrary1
Received response for /v2/codestructure but could not find request.
Received response for /v2/getcodeactions but could not find request.
```
</details>

Here is the repro: 
[CsWin32Sandbox.zip](https://github.com/microsoft/CsWin32/files/6167654/CsWin32Sandbox.zip)

**Reply by jasonmalinowski** [reply]
Alright will take a look next week.

**Reply by jasonmalinowski** [reply]
And by "next week" I clearly meant "a month or two". Sorry about that. 😢 

So sticking a debugger onto the Omnisharp process, it appears the generator is throwing an exception. We also seem to be eating the exception rather than reporting it, so we'll need to fix that! But the exception is:

    Could not load file or assembly 'System.Numerics.Vectors, Version=4.1.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

Which seems to point to the first problem: the generator package is packaging dependencies like System.Text.Json, but isn't packaging dependencies of System.Text.Json (namely System.Numerics.Vectors):

https://github.com/microsoft/CsWin32/blob/e67e9c55652afa1139409790622c1176e730db02/src/Microsoft.Windows.CsWin32/Microsoft.Windows.CsWin32.targets#L23-L35

So that probably should be updated to copy in the dependency. Why does this not matter for the command line compiler? Because command line Roslyn already ships System.Numerics.Vectors and has a binding redirect from the version CsWin32 is using to the version Roslyn ship.

I tried copying the version used into the NuGet package to test locally and found it still doesn't work. The second problem seems to be that Omnisharp isn't handling the dependencies like I'd expect it to. Omnisharp _does_ ship System.Numerics.Vectors, but doesn't have a binding redirect. In this case, I'd expect that the analyzer assembly loader should kick in, hook AssemblyResolve and provide the correct version. But if I look at [the Omnisharp's "did I hook resolve" flag](https://github.com/OmniSharp/omnisharp-roslyn/blob/d52c76d3218cf8c7be6cfe9ddb0539466b65745d/src/OmniSharp.Host/Services/AnalyzerAssemblyLoader.cs#L33) in the dump I see that's unset, which makes me think we also have a Omnisharp bug here as well. I'll file a bug momentarily for that.

I _can_ confirm if I hack in a binding redirect into the Omnisharp the generator then works.

**Reply by jasonmalinowski** [reply]
Ah, the Omnisharp code was copy/pasted from Roslyn, which had a bug: https://github.com/dotnet/roslyn/pull/53166

**Reply by AArnott** [reply]
Thanks. I filed #261 to track your suggested change in our package.

**Reply by jasonmalinowski** [reply]
And I've filed https://github.com/dotnet/roslyn/issues/53672 to track the next fix we need to make in Roslyn (and by extension Omnisharp) to make this work.

**Comment by BinToss** [comment]
I've decided to open an issue on OmniSharp/omnisharp-roslyn since the previous thread ended with several resolved Issues while seemingly not resolving this one.
https://github.com/OmniSharp/omnisharp-vscode/issues/5080
